src/main/java/htsjdk/samtools/CSIIndex.java

```diff
@@ -0,0 +1,466 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.seekablestream.SeekableStream;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.*;
+
+
+public class CSIIndex extends AbstractBAMFileIndex implements BrowseableBAMIndex {
+
+    private int binDepth;
+    private int minShift;
+    private int maxBins;
+    private int maxSpan;
+    private byte[] auxData;
+    private int nReferences;
+    private int metaDataPos = -1;
+
+    /**
+     * Constructors
+     */
+
+    public CSIIndex(final SeekableStream stream, final SAMSequenceDictionary dictionary) {
+        this(new IndexStreamBuffer(stream), stream.getSource(), dictionary);
+    }
+
+    public CSIIndex(final Path path, boolean enableMemoryMapping, final SAMSequenceDictionary dictionary) {
+        this(IndexFileBufferFactory.getBuffer(path.toFile(), enableMemoryMapping), path.toString(), dictionary);
+    }
+
+    public CSIIndex(final File file, boolean enableMemoryMapping, final SAMSequenceDictionary dictionary) {
+        this(IndexFileBufferFactory.getBuffer(file, enableMemoryMapping), file.getName(), dictionary);
+    }
+
+    private CSIIndex(final IndexFileBuffer indexFileBuffer, final String source, final SAMSequenceDictionary dictionary) {
+        super(indexFileBuffer, dictionary);
+
+        verifyCSIMagicNumber(source);
```

Comment:
I think we should probably make this a protected method on AbstractBamFileIndex and implement it there for bai and here for CSI.  Or we maybe make it abstract and pull out a common AbstractBaiFileIndex that is shared in common between the 2 bai implementations.  It would let us clean up the constructor chain in AbstractBamFileIndex so there isn't any constructor you can call that won't validate the magic number.